File: tutorial/05-arrays/problems/03-05-maximum-index.py
class Soft05p0305: 
    def solution(self,arr):
        length=len(arr)
        stack=[0]
        for i in range(1,length):
            if arr[stack[-1]] >= arr[i]:
                stack.append(i)
        mx=0 
        for i in range(length-1,-1,-1):
            while  stack and  arr[stack[-1]]<=arr[i]:
                 mx= max(mx,i-stack[-1])
                 stack.pop()
        return mx
soft05p0305= Soft05p0305()
print(soft05p0305.solution([34, 8, 10, 3, 2, 80, 30, 33, 1]))
    

File: tutorial/05-arrays/problems/03-05-maximum-index.ts
class Soft05P0305 {
  public static solution(arr: number[]) {
    const length = arr.length;
    const stack = [0];
    for (let i = 1; i < length; i++) {
      if (arr[i] <= arr[stack.length - 1]) {
        stack.push(i);
      }
    }
    let max = 0;
    for (let i = length - 1; i >= 0; i--) {
      while (stack.length > 0 && arr[i] >= arr[stack.length - 1]) {
        max = Math.max(max, i - stack[stack.length - 1]);
        stack.pop();
      }
    }
    return max;
  }
}
